### PR TITLE
Add proper comment offset support

### PR DESCRIPTION
**Pull request by nickvergessen**
The offset is based on the last known comment instead of limit-offset,
so new comments don't mess up requests which get the history of an object.

It also allows to sort in the other direction (to get comments newer than the given offset-comment)

"Forward port of https://github.com/nextcloud/spreed/pull/767"

### CODE REVIEW

**Review by MorrisJobke — APPROVED** [review]

**Comment by blizzz** [comment]
Somehow I don't see unit tests

File: lib/public/Comments/ICommentsManager.php [hunk 1]
```diff
@@ -120,6 +120,23 @@ public function getForObject(
...
+	 * @param string $objectType
+	 * @param string $objectId
+	 * @param int $lastKnownCommentId
+	 * @param string $sortDirection
+	 * @param int $limit
+	 * @return array
+	 * @since 14.0.0
+	 */
+	public function getForObjectSince(
+		$objectType,
```
Comment (rullzer) [inline lib/public/Comments/ICommentsManager.php]: Where are the types!!!

File: lib/private/Comments/Manager.php [hunk 2]
```diff
@@ -376,6 +376,121 @@ public function getForObject(
...
+		return $comments;
+	}
+
+	/**
+	 * @param string $objectType
+	 * @param string $objectId
+	 * @param int $id
+	 * @return Comment|null
+	 */
+	protected function getLastKnownComment($objectType,
```
Comment (rullzer) [inline lib/private/Comments/Manager.php]: types

File: lib/private/Comments/Manager.php [hunk 3]
```diff
@@ -376,6 +376,121 @@ public function getForObject(
...
 
+	/**
+	 * @param string $objectType
+	 * @param string $objectId
+	 * @param int $lastKnownCommentId
+	 * @param string $sortDirection
+	 * @param int $limit
+	 * @return array
+	 */
+	public function getForObjectSince(
```
Comment (rullzer) [inline lib/private/Comments/Manager.php]: types

**Review by rullzer — DISMISSED** [review]

File: lib/private/Comments/Manager.php (hunk 3 above)
Comment (nickvergessen) [reply lib/private/Comments/Manager.php]: fixed

**Comment by blizzz** [comment]
CI complains

File: lib/private/Comments/Manager.php [hunk 4]
```diff
@@ -376,6 +376,121 @@ public function getForObject(
...
 	}
 
+	/**
+	 * @param string $objectType
+	 * @param string $objectId
+	 * @param int $lastKnownCommentId
+	 * @param string $sortDirection
+	 * @param int $limit
+	 * @return array
+	 */
```
Comment (MorrisJobke) [inline lib/private/Comments/Manager.php]: No need to document the stuff that is already present as type hints. 😉

File: tests/lib/Comments/ManagerTest.php [hunk 5]
```diff
@@ -354,6 +354,48 @@ public function testGetNumberOfUnreadCommentsForFolder() {
 		], $amount);
 	}
 
+	/**
+	 * @dataProvider dataGetForObjectSince
+	 * @param $lastKnown
+	 * @param $order
+	 * @param $limit
+	 * @param $resultFrom
+	 * @param $resultTo
```
Comment (MorrisJobke) [inline tests/lib/Comments/ManagerTest.php]: Drop the PHPDoc for the params - see above.

File: lib/public/Comments/ICommentsManager.php [hunk 6]
```diff
@@ -120,6 +120,23 @@ public function getForObject(
 			\DateTime $notOlderThan = null
 	);
 
+	/**
+	 * @param string $objectType
+	 * @param string $objectId
+	 * @param int $lastKnownCommentId
+	 * @param string $sortDirection
+	 * @param int $limit
+	 * @return array
```
Comment (MorrisJobke) [inline lib/public/Comments/ICommentsManager.php]: No need to document the type hinted stuff. Best would be to describe the parameters (see next method documentation for example).

**Comment by nickvergessen** [comment]
Fixed

**Review by MorrisJobke — APPROVED** [review]

**Review by rullzer — APPROVED** [review]

**Pull request merged**